accepted/2021/workloads/workload-target-imports.md

```diff
@@ -0,0 +1,118 @@
+
+#### Exploring Import Order
+
+There are many `.props` and `.targets` files that get imported when building a .NET project. One way to see what is imported and in what order is by preprocessing a project, either with the `-pp:` [command line argument](https://docs.microsoft.com/en-us/visualstudio/msbuild/msbuild-command-line-reference) or with [MSBuild Structured Log Viewer](https://msbuildlog.com/). This will create a single aggregated project file with all project imports expanded inline.
+
+It is also possible to explore the project imports in the Visual Studio Solution Explorer. Clicking on the **Show All Files** button will add an **Imports** node to the solution tree under the project. You can expand this node and explore the tree of imports active in the project. You can also double click on an imported file to open it up in the editor and view its contents.
+
+![Project Imports in Solution Explorer](./solution-explorer-project-imports.png)
+
+#### Order of Imports in the .NET SDK
+
+The following is a simplified list of the files that are imported in an SDK-style .NET project:
+
+- `Directory.Build.props`
+- Main project file
+- Version logic (`Microsoft.NET.DefaultAssemblyInfo.targets`)
+- Output path logic (`Microsoft.NET.DefaultOutputPaths.targets`)
+- Publish profile
+- Target Framework parsing (`Microsoft.NET.TargetFrameworkInference.targets`)
+  - Also appends target framework to output and intermediate paths
+- Runtime identifier inference (`Microsoft.NET.RuntimeIdentifierInference.targets`)
+  - Also appends Runtime Identifier to output and intermediate paths
+- Workload targets imports
+- Language targets (e.g. `Microsoft.CSharp.targets`) and MSBuild common targets
+- `Directory.Build.targets`
+- (Rest of) .NET SDK targets
+- Old location for Windows, WindowsDesktop, and workload targets imports
+
+#### `Directory.Build.targets` import location
+
+Conceptually, `Directory.Build.targets` is imported after the body of the main project file. The exact location it is imported is not something most developers likely think about, but it is a good place to put common build logic that depends on properties set in the project file, such as the `TargetFramework`.
+
+There's not a perfect place to import `Directory.Build.targets`. However, given what we've learned, it may be that the best place to import it is after the TargetFramework parsing, and before the workloads are imported.  That way the logic in `Directory.Build.targets` would still be able to depend on the parsed TargetFramework information, but would not be able to override all of the targets and properties set by the .NET SDK and MSBuild common targets that it can today.
```

Comment:
Would it be at all possible to give users a bit more control over shimming each of these stages? E.g. in addition to Directory.Build.targets, maybe also Directory.Build.AfterSdks.targets and Directory.Build.BeforeSdk.targets. Sure, you'll still probably break things when you change things around, but then at least users that relied on the order would have a destination to move their definitions to.

Reply:
Also, very related to ordering, is there any chance that msbuild might eventually support lazily-evaluated properties? A large minority of the build ordering issues I help peers with revolves around reading properties too early (e.g. trying to read OutputPath in Directory.Build.props). Being able to set these using lazy evaluation (e.g. like gnu make `=` vs `:=`) in these cases would resolve a few issues. Admittedly, would create its own issues too, but worth a thought here.